File: src/CCVTAC.Console/Downloading/Downloader.cs
using CCVTAC.Console.ExternalTools;
using MediaTypeWithUrls = CCVTAC.FSharp.Downloading.MediaType;
using UserSettings = CCVTAC.FSharp.Settings.UserSettings;

namespace CCVTAC.Console.Downloading;

internal static class Downloader
{
    internal static readonly string ProgramName = "yt-dlp";

    private record Urls(string Primary, string? Supplementary);

    internal static Result<MediaTypeWithUrls> WrapUrlInMediaType(string url)
    {
        var result = FSharp.Downloading.MediaTypeWithIds(url);

        return result.IsOk ? Result.Ok(result.ResultValue) : Result.Fail(result.ErrorValue);
    }

    /// <summary>
    /// Completes the actual download process.
    /// </summary>
    /// <returns>A `Result` that, if successful, contains the name of the successfully downloaded format.</returns>
    internal static Result<string?> Run(
        MediaTypeWithUrls mediaType,
        UserSettings settings,
        Printer printer
    )
    {
        if (mediaType is { IsVideo: false, IsPlaylistVideo: false })
        {
            printer.Info("Please wait for multiple videos to be downloaded...");
        }

        var rawUrls = FSharp.Downloading.ExtractDownloadUrls(mediaType);
        var urls = new Urls(rawUrls[0], rawUrls.Length == 2 ? rawUrls[1] : null);

        var downloadResult = new Result<(int, string)>();
        string? successfulFormat = null;

        foreach (string format in settings.AudioFormats)
        {
            string args = GenerateDownloadArgs(format, settings, mediaType, urls.Primary);
            string commandWithArgs = $"{ProgramName} {args}";
            var downloadSettings = new ToolSettings(commandWithArgs, settings.WorkingDirectory!);

            downloadResult = Runner.Run(downloadSettings, otherSuccessExitCodes: [1], printer);

            if (downloadResult.IsSuccess)
            {
                successfulFormat = format;

                var (exitCode, warnings) = downloadResult.Value;
                if (exitCode != 0)
                {
                    printer.Warning("Downloading completed with minor issues.");
                    if (warnings.HasText())
                    {
                        printer.Warning(warnings);
                    }
                }

                break;
            }

            printer.Debug($"Failure downloading \"{format}\" format.");
        }

        var errors = downloadResult.Errors.Select(e => e.Message).ToList();

        int audioFileCount = IoUtilities.Directories.AudioFileCount(settings.WorkingDirectory);
        if (audioFileCount == 0)
        {
            return Result.Fail(
                string.Join(Environment.NewLine, errors.Prepend("No audio files were downloaded."))
            );
        }

        if (errors.Count != 0)
        {
            downloadResult.Errors.ForEach(e => printer.Error(e.Message));
            printer.Info("Post-processing will still be attempted."); // For any partial downloads
        }
        else if (urls.Supplementary is not null)
        {
            string supplementaryArgs = GenerateDownloadArgs(
                null,
                settings,
                null,
                urls.Supplementary
            );

            var commandWithArgs = $"{ProgramName} {supplementaryArgs}";

            var supplementaryDownloadSettings = new ToolSettings(commandWithArgs, settings.WorkingDirectory!);

            var supplementaryDownloadResult = Runner.Run(
                supplementaryDownloadSettings,
                otherSuccessExitCodes: [1],
                printer
            );

            if (supplementaryDownloadResult.IsSuccess)
            {
                printer.Info("Supplementary download completed OK.");
            }
            else
            {
                printer.Error("Supplementary download failed.");
                errors.AddRange(supplementaryDownloadResult.Errors.Select(e => e.Message));
            }
        }

        return errors.Count > 0
            ? Result.Fail(string.Join(" / ", errors))
            : Result.Ok(successfulFormat);
    }

    /// <summary>
    /// Generate the entire argument string for the download tool.
    /// </summary>
    /// <param name="audioFormat">One of the supported audio format codes.</param>
    /// <param name="settings"></param>
    /// <param name="mediaType">A `MediaType` or null (which indicates a metadata-only supplementary download).</param>
    /// <param name="additionalArgs"></param>
    /// <returns>A string of arguments that can be passed directly to the download tool.</returns>
    private static string GenerateDownloadArgs(
        string? audioFormat,
        UserSettings settings,
        MediaTypeWithUrls? mediaType,
        params string[]? additionalArgs
    )
    {
        const string writeJson = "--write-info-json";
        const string trimFileNames = "--trim-filenames 250";

        // yt-dlp warning: "-f best" selects the best pre-merged format which is often not the best option.
        // To let yt-dlp download and merge the best available formats, simply do not pass any format selection."
        var formatArg =
            !audioFormat.HasText() || audioFormat == "best" ? string.Empty : $"-f {audioFormat}";

        HashSet<string> args = mediaType switch
        {
            // For metadata-only downloads
            null => [$"--flat-playlist {writeJson} {trimFileNames}"],

            // For video(s) with their respective metadata files (JSON and artwork).
            _ =>
            [
                "--extract-audio",
                formatArg,
                $"--audio-quality {settings.AudioQuality}",
                "--write-thumbnail --convert-thumbnails jpg", // For album art
                writeJson, // Contains metadata
                trimFileNames,
                "--retries 2", // Default is 10, which seems like overkill
            ],
        };

        // yt-dlp has a `--verbose` option too, but that's too much data.
        // It might be worth incorporating it in the future as a third option.
        args.Add(settings.QuietMode ? "--quiet --no-warnings" : string.Empty);

        if (mediaType is not null)
        {
            if (settings.SplitChapters)
            {
                args.Add("--split-chapters");
            }

            if (mediaType is { IsVideo: false, IsPlaylistVideo: false })
            {
                args.Add($"--sleep-interval {settings.SleepSecondsBetweenDownloads}");
            }

            // The numbering of regular playlists should be reversed because the newest items are
            // always placed at the top of the list at position #1. Instead, the oldest items
            // (at the end of the list) should begin at #1.
            if (mediaType.IsStandardPlaylist)
            {
                // The digits followed by `B` induce trimming to the specified number of bytes.
                // Use `s` instead of `B` to trim to a specified number of characters.
                // Reference: https://github.com/yt-dlp/yt-dlp/issues/1136#issuecomment-1114252397
                // Also, it's possible this trimming should be applied to `ReleasePlaylist`s too.
                args.Add(
                    """-o "%(uploader).80B - %(playlist).80B - %(playlist_autonumber)s - %(title).150B [%(id)s].%(ext)s" --playlist-reverse"""
                );
            }
        }

        return string.Join(" ", args.Concat(additionalArgs ?? []));
    }
}

File: src/CCVTAC.FSharp/Downloading/Downloader.fs
namespace CCVTAC.Console.Downloading

open CCVTAC.Console
open CCVTAC.Console.ExternalTools.Runner
open CCVTAC.Console.IoUtilities
open CCVTAC.Console.IoUtilities.Directories
open CCVTAC.Console.Downloading.Downloading
open CCVTAC.Console.ExternalTools
open CCVTAC.Console.Settings.Settings
open FsToolkit.ErrorHandling
open System

module Downloader =

    [<Literal>]
    let private programName = "yt-dlp"

    // TODO: Is the audioFormat not in the settings?
    /// Generate the entire argument string for the download tool.
    /// audioFormat: one of the supported audio format codes (or null for none)
    /// mediaType: Some MediaType for normal downloads, None for metadata-only supplementary downloads
    /// additionalArgs: optional extra args (e.g., the URL)
    let generateDownloadArgs audioFormat userSettings (mediaType: MediaType option) additionalArgs : string =
        let writeJsonArg = "--write-info-json"
        let trimFileNamesArg = "--trim-filenames 250"

        let formatArg =
            match audioFormat with
            | None -> String.Empty
            | Some f when f = "best" -> String.Empty
            | Some f -> $"-f {f}"

        let mutable args =
            match mediaType with
            | None ->
                [ $"--flat-playlist {writeJsonArg} {trimFileNamesArg}" ]
            | Some _ ->
                [ $"--extract-audio {formatArg}"
                  $"--audio-quality {userSettings.AudioQuality}"
                  "--write-thumbnail --convert-thumbnails jpg"
                  writeJsonArg
                  trimFileNamesArg
                  "--retries 2" ]
            |> Set.ofList

        if userSettings.QuietMode then
            args <- args.Add "--quiet --no-warnings"

        // No MediaType indicates that this is a supplemental metadata-only download.
        // TODO: Add a union type to more clearly indicate this difference.
        match mediaType with
        | Some mt ->
            if userSettings.SplitChapters then
                args <- args.Add "--split-chapters"

            if not mt.IsVideo && not mt.IsPlaylistVideo then
                args <- args.Add $"--sleep-interval {userSettings.SleepSecondsBetweenDownloads}"

            if mt.IsStandardPlaylist then
                args <- args.Add
                    """-o "%(uploader).80B - %(playlist).80B - %(playlist_autonumber)s - %(title).150B [%(id)s].%(ext)s" --playlist-reverse"""
        | None -> ()

        let extraArgs = defaultArg additionalArgs [] |> Set.ofList
        String.Join(" ", Set.union args extraArgs)

    let wrapUrlInMediaType url : Result<MediaType, string> =
        mediaTypeWithIds url

    /// Completes the actual download process.
    let downloadMedia (printer: Printer) (mediaType: MediaType) userSettings (PrimaryUrl url)
        : Result<string list, string list> =

        if not mediaType.IsVideo && not mediaType.IsPlaylistVideo then
            printer.Info("Please wait for multiple videos to be downloaded...")

        let rec loop errors audioFormats =
            match audioFormats with
            | [] ->
                Error errors
            | format :: formats ->
                let args = generateDownloadArgs (Some format) userSettings (Some mediaType) (Some [url])
                let commandWithArgs = $"{programName} {args}"
                let downloadSettings = ToolSettings.create commandWithArgs userSettings.WorkingDirectory

                let downloadResult = runTool downloadSettings [1] printer
                let filesDownloaded = audioFileCount userSettings.WorkingDirectory Files.audioFileExtensions > 0

                match downloadResult, filesDownloaded with
                | Ok result, true ->
                    Ok <|
                        $"Successfully downloaded the \"{format}\" format."
                        :: match result.Error with
                            | Some err -> [$"However, a minor issue was reported: {err}"]
                            | None -> []
                | Ok result, false ->
                    Error <|
                        $"The \"{format}\" format download was reported as successful, but no audio files were downloaded!"
                        :: match result.Error with
                            | Some err -> [$"Error: {err}"]
                            | None -> []
                | Error err, true ->
                    Error <|
                        [$"The downloader reported failure for \"{format}\", yet audio files were unexpectedly downloaded!"
                         $"Error: {err}"]
                | Error err, false ->
                    let newErr = $"A download error was reported for the \"{format}\" format, and no audio files were downloaded. {err}"
                    loop (List.append errors [newErr]) formats

        loop [] userSettings.AudioFormats

    let downloadMetadata (printer: Printer) userSettings (SupplementaryUrl url)
        : Result<string list, string list> =

        match url with
        | None -> Ok ["No supplementary metadata link found."]
        | Some url' ->
            let args = generateDownloadArgs None userSettings None (Some [url'])
            let commandWithArgs = $"{programName} {args}"
            let downloadSettings = ToolSettings.create commandWithArgs userSettings.WorkingDirectory
            let metadataDownloadResult = runTool downloadSettings [1] printer

            match metadataDownloadResult with
            | Ok _ -> Error ["Supplementary metadata download completed OK."]
            | Error err -> Error [$"Supplementary metadata download failed: {err}"]

    /// Undertakes the actual download process.
    let run (mediaType: MediaType) userSettings (printer: Printer) : Result<string list, string list> =
        result {
            let rawUrls = generateDownloadUrl mediaType
            let urls =
                { Primary = PrimaryUrl rawUrls[0]
                  Metadata = SupplementaryUrl <| if rawUrls.Length = 2 then Some rawUrls[1] else None }
            let! _ = downloadMedia printer mediaType userSettings urls.Primary
            let! metadataDownloadResult = downloadMetadata printer userSettings urls.Metadata
            return! Ok metadataDownloadResult
        }
